data-analytics/iot-nirvana/pipeline/src/test/java/com/google/cloud/demo/iot/nirvana/pipeline/FilterMessagesTest.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.demo.iot.nirvana.pipeline;
+
+import com.google.cloud.demo.iot.nirvana.common.CityTemperature;
+import com.google.cloud.demo.iot.nirvana.common.Message;
+import com.google.cloud.demo.iot.nirvana.common.TemperatureUtils;
+import com.google.gson.Gson;
+import java.io.Serializable;
+import java.io.UnsupportedEncodingException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.testing.NeedsRunner;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link FilterMessages}. */
+@RunWith(JUnit4.class)
+public final class FilterMessagesTest implements Serializable {
+
+  @Rule public final transient TestPipeline pipeline = TestPipeline.create();
+
+  private String invalidMessage;
+  private String validMessageTiranaStr;
+  private String validMessageOranjestadStr;
+  private Message validMessageTirana;
+  private Message validMessageOranjestad;
+  private CityTemperature cityTirana;
+  private CityTemperature cityOranjestad;
+  private PCollectionView<Map<String, CityTemperature>> cities;
+  private TupleTag<KV<Message, CityTemperature>> goodTag;
+  private TupleTag<String> errorTag;
+
+  /** Create test data. */
+  @Before
+  public void createTestData() throws NoSuchAlgorithmException, UnsupportedEncodingException {
```

Comment:
A lot of these can be static final in my opinion.

Reply:
Moved TupleTags to static final in commit d875a04afcbd40c6b2f1462fe1f5370e94abf34c. The other test data class attributes cannot be final.